File: Examples/Getting-Started/Create-Device/Python/netld_example_client.py
import json
import os
import uuid

import requests
from dotenv import load_dotenv


class NetLDError(RuntimeError):
    pass


def load_environment(path):
    load_dotenv(path, override=True)


class NetLDClient:
    def __init__(self, base_url, api_key, timeout=10):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()

    @classmethod
    def from_env(cls):
        base_url = os.getenv("NETLD_BASE_URL", "").strip()
        api_key = os.getenv("NETLD_API_KEY", "").strip()
        if not base_url:
            raise NetLDError("Set NETLD_BASE_URL in .env before running this example.")
        if not api_key:
            raise NetLDError("Set NETLD_API_KEY in .env before running this example.")
        return cls(base_url, api_key)

    @property
    def headers(self):
        return {"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"}

    def login(self):
        try:
            response = self.session.get(
                f"{self.base_url}/rest",
                headers={"Authorization": f"Bearer {self.api_key}"},
                timeout=self.timeout,
                verify=True,
                allow_redirects=False,
            )
        except requests.RequestException as exc:
            raise NetLDError(f"Could not reach {self.base_url}.") from exc
        if response.is_redirect:
            raise NetLDError(f"Login redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise NetLDError(f"Login failed with HTTP {response.status_code}.") from exc

    def call(self, method, **parameters):
        payload = {"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())}
        try:
            response = self.session.post(
                f"{self.base_url}/rest",
                headers=self.headers,
                json=payload,
                timeout=self.timeout,
                verify=True,
                allow_redirects=False,
            )
        except requests.RequestException as exc:
            raise NetLDError(f"Could not reach {self.base_url}.") from exc
        if response.is_redirect:
            raise NetLDError(f"API call redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise NetLDError(f"API call failed with HTTP {response.status_code}.") from exc
        data = response.json()
        if data.get("error"):
            raise NetLDError(json.dumps(data["error"]))
        return data.get("result")

    def get_device(self, network, ip_address):
        return self.call("Inventory.getDevice", network=network, ipAddress=ip_address)

    def create_device(self, network, ip_address, adapter_id):
        return self.call(
            "Inventory.createDevice", network=network, ipAddress=ip_address, adapterId=adapter_id
        )

File: Examples/Getting-Started/Create-Device/nodeJS/netld-example-client.mjs
import { randomUUID } from "node:crypto";

export class NetLDError extends Error {}

export class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 10000) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeout = timeout;
    this.cookieHeader = "";
  }

  headers(extra = {}) {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      ...(this.cookieHeader ? { Cookie: this.cookieHeader } : {}),
      ...extra,
    };
  }

  updateCookies(response) {
    const values = response.headers.getSetCookie?.() || splitSetCookie(response.headers.get("set-cookie") || "");
    this.cookieHeader = values.map((value) => value.split(";", 1)[0]).filter(Boolean).join("; ");
  }

  async login() {
    const response = await this.request("/rest");
    this.updateCookies(response);
  }

  async request(path, options = {}) {
    let response;
    try {
      response = await fetch(`${this.baseUrl}${path}`, {
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
        ...options,
        headers: this.headers(options.headers),
      });
    } catch {
      throw new NetLDError(`Could not reach ${this.baseUrl}.`);
    }
    if (response.status >= 300 && response.status < 400) {
      throw new NetLDError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new NetLDError(`Request failed with HTTP ${response.status}.`);
    return response;
  }

  async call(method, parameters = {}) {
    const response = await this.request("/rest", {
      method: "POST",
      body: JSON.stringify({ jsonrpc: "2.0", method, params: parameters, id: randomUUID() }),
    });
    const data = await response.json();
    if (data.error) throw new NetLDError(JSON.stringify(data.error));
    return data.result;
  }

  getDevice({ network, ipAddress }) {
    return this.call("Inventory.getDevice", { network, ipAddress });
  }

  createDevice({ network, ipAddress, adapterId }) {
    return this.call("Inventory.createDevice", { network, ipAddress, adapterId });
  }
}

function splitSetCookie(header) {
  return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : [];
}

